File: fireplace-app/src/main/java/io/github/bric3/fireplace/charts/LineChartRenderer.java
package io.github.bric3.fireplace.charts;

import java.awt.*;
import java.awt.geom.AffineTransform;
import java.awt.geom.Ellipse2D;
import java.awt.geom.Path2D;
import java.awt.geom.Rectangle2D;
import java.util.Objects;

/**
 * A renderer that draws line charts.
 */
public class LineChartRenderer extends ChartRenderer {

    /**
     * The stroke for rendering the line.
     */
    private final Stroke lineStroke;

    /**
     * The paint for rendering the line.
     */
    private final Paint linePaint;

    /**
     * The fill color - if null, then the line is drawn without a fill.
     */
    private Color fillColor;

    /**
     * The gradient fill colors - an alternative to the fill color, this pair of
     * colors will be used to create a vertical gradient fill.  If null, no fill
     * will be drawn.
     */
    private Color[] gradientFillColors;

    /**
     * A shape that will be drawn at the point of the last value in the series.
     * Leave as null if you don't want anything drawn.
     */
    private final Shape lastValueShape;

    /**
     * The color for the shape drawn at the location of the most recent value.
     */
    private final Color lastValueShapeColor;

    /**
     * A flag that controls whether zero must be included in the y-value range.
     */
    private final boolean includeZeroInYRange;

    /**
     * Creates a new renderer with default attributes.
     */
    public LineChartRenderer() {
        this(new BasicStroke(1.0f, BasicStroke.CAP_ROUND, BasicStroke.JOIN_ROUND), Color.BLACK);
    }

    /**
     * Creates a new renderer that draws a line with the specified stroke and paint.
     *
     * @param stroke the line stroke ({@code null} not permitted).
     * @param paint  the line paint ({@code null} not permitted).
     */
    public LineChartRenderer(Stroke stroke, Paint paint) {
        Objects.requireNonNull(stroke);
        Objects.requireNonNull(paint);
        this.lineStroke = stroke;
        this.linePaint = paint;
        this.fillColor = null; // no fill
        this.gradientFillColors = null; // no gradient fill
        this.lastValueShape = new Ellipse2D.Double(-2, -2, 4, 4);
        this.lastValueShapeColor = new Color(25, 90, 25);
        this.includeZeroInYRange = true;
    }

    /**
     * Sets the color used to fill the area under the line.  If {@code null}, the area will not be filled.
     *
     * @param color the new fill color ({@code null} permitted).
     */
    public void setFillColor(Color color) {
        this.fillColor = color;
    }

    /**
     * Sets the gradient fill colors.
     *
     * @param colors the top color ({@code null} permitted).
     */
    public void setGradientFillColors(Color[] colors) {
        this.gradientFillColors = colors;
    }

    /**
     * Draws a line chart within the specified bounds.
     *
     * @param chart      the chart being drawn.
     * @param dataset    the dataset.
     * @param g2         the Java2D graphics target.
     * @param plotBounds the bounds within which the data should be rendered.
     */
    @Override
    public void draw(Chart chart, XYDataset dataset, Graphics2D g2, Rectangle2D plotBounds) {
        super.draw(chart, dataset, g2, plotBounds);

        // within the bounds, draw a line chart
        int itemCount = dataset.getItemCount();
        if (itemCount == 0) {
            // what to do? FIXME
            throw new IllegalArgumentException("No data");
        }
        var xRange = dataset.rangeOfX;
        var yRange = dataset.rangeOfY;
        if (includeZeroInYRange) {
            yRange = yRange.include(0.0);
        }
        // TODO - what if the yRange is a single value or all values are null?

        // does the chart have a fixed display range?
        // PLUS are there range constraints on the chart itself (for example, y in the range 0 to 100)
        Path2D path = new Path2D.Double();
        for (int i = 0; i < itemCount; i++) {
            if (yRange == null || yRange.isZeroLength()) continue;
            double xx = plotBounds.getX() + xRange.calcFraction(dataset.xAt(i)) * plotBounds.getWidth();
            if (dataset.yAt(i) == null) continue;
            double yy = plotBounds.getMaxY() - yRange.calcFraction(dataset.yAt(i)) * plotBounds.getHeight();
            if (i == 0) {
                path.moveTo(xx, yy);
            } else {
                path.lineTo(xx, yy);
            }
        }

        // optionally the area under the line can be filled - do this first...
        if (fillColor != null || gradientFillColors != null) {
            Path2D path2 = new Path2D.Double(path);
            path2.lineTo(path.getCurrentPoint().getX(), plotBounds.getMaxY());
            path2.lineTo(plotBounds.getX(), plotBounds.getMaxY());
            path2.closePath();
            g2.setPaint(Objects.requireNonNullElseGet(
                    fillColor,
                    () -> verticalGradientPaint(gradientFillColors, plotBounds)
            ));
            g2.fill(path2);
        }

        g2.setPaint(linePaint);
        g2.setStroke(lineStroke);
        g2.draw(path);

        if (this.lastValueShape != null) {
            AffineTransform saved = g2.getTransform();
            g2.translate(path.getCurrentPoint().getX(), path.getCurrentPoint().getY());
            g2.setColor(this.lastValueShapeColor);
            g2.fill(this.lastValueShape);
            g2.setTransform(saved);
        }
    }

    /**
     * Creates a new {@link GradientPaint} instance based on the supplied colors and with coordinates
     * set to run the gradient from the top to the bottom of the supplied bounds.
     *
     * @param gradientColors the top color ({@code null} not permitted).
     * @param bounds         the bounds ({@code null} not permitted).
     * @return A GradientPaint.
     */
    public static GradientPaint verticalGradientPaint(Color[] gradientColors, Rectangle2D bounds) {
        if (gradientColors == null || gradientColors.length != 2 || gradientColors[0] == null || gradientColors[1] == null) {
            throw new IllegalArgumentException("gradientColors must be non-null and have 2 colors");
        }
        return new GradientPaint(
                (float) bounds.getCenterX(), (float) bounds.getY(), gradientColors[0], // top
                (float) bounds.getCenterX(), (float) bounds.getMaxY(), gradientColors[1] // bottom
        );
    }
}

File: fireplace-app/src/main/kotlin/io/github/bric3/fireplace/charts/LineChartRenderer.kt
/*
 * Fireplace
 *
 * Copyright (c) 2021, Today - Brice Dutheil
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package io.github.bric3.fireplace.charts

import io.github.bric3.fireplace.core.ui.Colors
import io.github.bric3.fireplace.core.ui.LightDarkColor
import java.awt.BasicStroke
import java.awt.Color
import java.awt.GradientPaint
import java.awt.Graphics2D
import java.awt.Paint
import java.awt.Point
import java.awt.Stroke
import java.awt.geom.Ellipse2D
import java.awt.geom.Path2D
import java.awt.geom.Rectangle2D
import kotlin.math.abs


/**
 * A renderer that draws line charts.
 *
 * @param stroke the line stroke.
 * @param paint  the line paint.
 */
class LineChartRenderer(
    stroke: Stroke = BasicStroke(
        1.2f,
        BasicStroke.CAP_ROUND,
        BasicStroke.JOIN_ROUND
    ),
    paint: Paint = LightDarkColor(Color.BLACK, Color.WHITE),
) : ChartRenderer {

    /**
     * The gradient fill colors, either `null`, one color, or at most two colors.
     * When two colors are passed, this pair of colors will be used to create
     * a vertical gradient fill.
     * If `null`, no fill will be drawn.
     */
    var fillColors: List<Color>? = null

    /**
     * The stroke for rendering the line.
     */
    var lineStroke = stroke

    /**
     * The paint for rendering the line.
     */
    var linePaint = paint

    /**
     * A shape that will be drawn at the point of the last value in the series.
     * Leave as null if you don't want anything drawn.
     */
    private val lastValueShape = Ellipse2D.Double(-2.0, -2.0, 4.0, 4.0)

    /**
     * The color for the shape drawn at the location of the most recent value.
     */
    private val lastValueShapeColor = Color(25, 90, 25)

    /**
     * A flag that controls whether zero value must be included in the y-value range.
     */
    private val includeZeroInYRange = true

    override fun drawTrackerLine(
        chart: Chart,
        dataset: ChartDataset,
        g2: Graphics2D,
        plotBounds: Rectangle2D,
        mousePosition: Point?
    ) {
        val itemCount = dataset.itemCount
        if (itemCount == 0) return;
        val xRange = dataset.rangeOfX
        var yRange = dataset.rangeOfY
        if (includeZeroInYRange) {
            yRange = yRange.include(0.0)
        }

        val showTracker = mousePosition != null && mousePosition.x >= plotBounds.x && mousePosition.x <= plotBounds.maxX
        if (!showTracker) return

        var closestItemIndex = 0
        var closestItemX = Double.MAX_VALUE
        for (i in 0 until itemCount) {
            if (yRange.isZeroLength) continue
            val xx = plotBounds.x + xRange.ratioFor(dataset.xAt(i)) * plotBounds.width
            // discover the closest item to the mouse adjusted position
            val distance = abs(xx - mousePosition!!.x)
            if (distance < closestItemX) {
                closestItemX = distance
                closestItemIndex = i
            }
        }

        // draw vertical line
        val centerX = plotBounds.x + xRange.ratioFor(dataset.xAt(closestItemIndex)) * plotBounds.width
        val centerY = plotBounds.maxY - yRange.ratioFor(dataset.yAt(closestItemIndex)!!) * plotBounds.height

        g2.paint = Color.GRAY // TODO get color from chart
        g2.drawLine(
            centerX.toInt(),
            plotBounds.y.toInt(),
            centerX.toInt(),
            plotBounds.maxY.toInt()
        )
    }

    /**
     * Draws a line chart within the specified bounds.
     *
     * @param chart         the chart being drawn.
     * @param dataset       the dataset.
     * @param g2            the Java2D graphics target.
     * @param plotBounds    the bounds within which the data should be rendered.
     * @param mousePosition the mouse position.
     */
    override fun draw(
        chart: Chart,
        dataset: ChartDataset,
        g2: Graphics2D,
        plotBounds: Rectangle2D,
        mousePosition: Point?
    ) {
        // within the bounds, draw a line chart
        val itemCount = dataset.itemCount
        if (itemCount == 0) return;
        val xRange = dataset.rangeOfX
        var yRange = dataset.rangeOfY
        if (includeZeroInYRange) {
            yRange = yRange.include(0.0)
        }

        val showTracker = mousePosition != null && mousePosition.x >= plotBounds.x && mousePosition.x <= plotBounds.maxX

        var closestItemIndex = 0
        var closestItemX = Double.MAX_VALUE

        // TODO - what if the yRange is a single value or all values are null?
        // Does the chart have a fixed display range?
        // Also, are there range constraints on the chart itself
        // (for example, y in the range 0 to 100)
        val path: Path2D = Path2D.Double()
        for (i in 0 until itemCount) {
            if (yRange.isZeroLength) continue
            val xx = plotBounds.x + xRange.ratioFor(dataset.xAt(i)) * plotBounds.width
            // discover the closest item to the mouse adjusted position
            if (showTracker) {
                val distance = abs(xx - mousePosition!!.x)
                if (distance < closestItemX) {
                    closestItemX = distance
                    closestItemIndex = i
                }
            }

            val yValue = dataset.yAt(i) ?: continue
            val yy = plotBounds.maxY - yRange.ratioFor(yValue) * plotBounds.height
            if (i == 0) {
                path.moveTo(xx, yy)
            } else {
                path.lineTo(xx, yy)
            }
        }

        // Optionally, the area under the line can be filled - do this first...
        fillColors?.let {
            val path2 = Path2D.Double(path).apply {
                lineTo(path.currentPoint.x, plotBounds.maxY)
                lineTo(plotBounds.x, plotBounds.maxY)
                closePath()
            }
            g2.paint = it.singleOrNull() ?: verticalGradientPaint(it, plotBounds)
            g2.fill(path2)
        }


        g2.paint = linePaint
        g2.stroke = lineStroke
        g2.draw(path)

        // paint last value
        run {
            val saved = g2.transform
            g2.translate(path.currentPoint.x, path.currentPoint.y)
            g2.color = lastValueShapeColor
            g2.fill(this.lastValueShape)
            g2.transform = saved
        }

        if (showTracker) {
            val centerX = plotBounds.x + xRange.ratioFor(dataset.xAt(closestItemIndex)) * plotBounds.width
            val centerY = plotBounds.maxY - yRange.ratioFor(dataset.yAt(closestItemIndex)!!) * plotBounds.height
            
            val dotRadius = 5.0
            val dot = Ellipse2D.Double(
                centerX - dotRadius,
                centerY - dotRadius,
                2.0 * dotRadius,
                2.0 * dotRadius
            )

            val dotBorderRadius = 5.0
            val dotBorder = Ellipse2D.Double(
                centerX - dotBorderRadius,
                centerY - dotBorderRadius,
                2.0 * dotBorderRadius,
                2.0 * dotBorderRadius
            )

            g2.paint = linePaint
            g2.fill(dot)

            g2.paint = Colors.panelBackground
            g2.draw(dotBorder)
        }
    }

    companion object {
        /**
         * Creates a new [GradientPaint] instance based on the supplied colors and with coordinates
         * set to run the gradient from the top to the bottom of the supplied bounds.
         *
         * @param gradientColors the top color (`null` not permitted).
         * @param bounds         the bounds (`null` not permitted).
         * @return A GradientPaint.
         */
        fun verticalGradientPaint(gradientColors: List<Color>, bounds: Rectangle2D): GradientPaint {
            require(gradientColors.size == 2) { "gradientColors must be non-null and have 2 colors" }
            return GradientPaint(
                bounds.centerX.toFloat(), bounds.y.toFloat(), gradientColors[0],   // top
                bounds.centerX.toFloat(), bounds.maxY.toFloat(), gradientColors[1] // bottom
            )
        }
    }
}
